src/metabase/query_processor/middleware/constraints.clj

```diff
@@ -5,33 +5,46 @@
    [metabase.models.setting :as setting]
    [metabase.util.i18n :refer [deferred-tru]]))
 
-(def ^:private ^:const default-max-results-bare-rows 2000)
+(def ^:private ^:const default-max-unaggregated-query-row-limit 2000)
+(def ^:private ^:const default-max-aggregated-query-row-limit 10000)
 
-;; NOTE: this was changed from a hardcoded var with value of 2000 (now moved to [[default-max-results-bare-rows]])
+;; NOTE: this was changed from a hardcoded var with value of 2000 (now moved to [[default-max-unaggregated-query-row-limit]])
 ;; to a setting in 0.43 the setting, which allows for DB local value, can still be nil, so any places below that used
 ;; to reference the former constant value have to expect it could return nil instead
-(setting/defsetting max-results-bare-rows
+(setting/defsetting max-unaggregated-query-row-limit
   (deferred-tru "Maximum number of rows to return specifically on :rows type queries via the API.")
   :visibility     :authenticated
   :type           :integer
   :database-local :allowed
   :audit          :getter)
 
-(def ^:private max-results
-  "General maximum number of rows to return from an API query."
-  10000)
+(setting/defsetting max-aggregated-query-row-limit
+  (deferred-tru "Maximum number of rows to return for aggregated queries via the API.")
+  :visibility     :authenticated
+  :type           :integer
+  :database-local :allowed
+  :audit          :getter)
```

Comment:
It is confusing - I'll add comments to explain. This could possibly use a refactor, but we want to distinguish between:

- the user **HAS** set a value and the constraints middleware **WAS** applied: we use the specified value
- the user **HAS NOT** set a value and the constraints middleware **WAS** applied: we use the "default" value
- the user **HAS** set a value and the constraints middleware **WAS NOT** applied: we use the specified value
- the user **HAS NOT** set a value and the constraints middleware **WAS NOT** applied: we use the "absolute max results" value

The current code works by looking for, in order:
1. a non-nil value set by the constraints **middleware** (so the value of the setting, the "default", or `nil` if no constraint middleware)
2. a non-nil value set by the constraints **setting** (so the value of the setting or `nil` if it wasn't set)
3. the "absolute max results"

I didn't want to mess too much with QP code, so I left the logic pretty much as it was. The only real change was which setting we're looking at in step 2.

Reply:
I added a comment [here](https://github.com/metabase/metabase/pull/35600/files#diff-3b2ac9a54646c3fc99a8239c4ae27cbc1478f01837a6e423c6a8b96a1a32c6e5R8-R22) with a better explanation for why we aren't applying defaults to the settings themselves.